# Credit to IzarLab shared script by Dr Tagore
library(numbat)
library(dplyr)
library(Seurat)
library(ggplot2)
library(glue)
library(data.table)
library(ggtree)
library(stringr)
library(tidygraph)
library(patchwork)
library(viridis)

'%notin%' <- Negate('%in%')
message_ <- function(m) {
    message(paste(Sys.time(), ': ', m))
}

# Provide sample name as argument

# Read in complete cohort

args <- commandArgs(trailingOnly=TRUE)
pat <- args[1] # 'GM18'

projectdir <- file.path('/home/ubuntu/data/cxcr4-pdac')
seu_file <- file.path(projectdir, 'seurat', pat, paste0(pat, '_cb.rds'))
allele_pileup <- file.path(projectdir, 'numbat', pat, paste0(pat, '_allele_counts.tsv.gz'))
outdir <- file.path(projectdir, 'numbat', pat, 'numbat_downstream')
numbat_out_rds <- file.path(outdir, paste0(pat, '_numbat_out.rds'))
cnv_calls_file <- file.path(outdir, paste0(pat,'_numbat_cnv_calls.csv'))
plots_file <- file.path(outdir, paste0(pat,'_numbat.pdf'))
numbat_clone_file <- file.path(outdir, paste0(pat,'_numbat_clones.rds'))
numbat_seu_file <- file.path(outdir, paste0(pat,'_numbat_seu.rds'))
numbat_seu_metadata_file <- file.path(outdir, paste0(pat,'_numbat_seu_metadata.csv'))

message_(paste('Processing:', pat))

if (!file.exists(numbat_out_rds)) {
    seu <- readRDS(seu_file)
    notum <- readRDS('/home/ubuntu/data/cxcr4-pdac/seurat/reference.RDS') # subset(seu,celltype_bped_main%in%c('Fibroblasts')) 

    # Get count_mats
    count_mat<-seu@assays$RNA@counts
    #colnames(count_mat)<-seu$barcode_orig
    ref_mat<-notum@assays$RNA@counts
    #colnames(ref_mat)<-notum$barcode_orig

    # count_mat is a gene x cell raw count matrices
    # cell_annot is a dataframe with columns "cell" and "group"
    ###cell_annot=data.frame(cell=colnames(ref_mat),group=notum@meta.data$manual.annot)
    cell_annot=data.frame(cell=colnames(ref_mat),group=notum@meta.data$celltype_bped_main)
    #cell_annot=data.frame(cell=notum$barcode_orig,group=notum$annot)
    ref_internal = aggregate_counts(ref_mat, cell_annot)

    # allele dataframe generated by pileup_and_phase script
    df_allele <- read.delim(allele_pileup)
    #df_allele<- subset(df_allele,cell %in% seu$barcode_orig)
    df_allele <- subset(df_allele,cell %in% colnames(count_mat))  # The barcodes that made the final cut

    #rm(list = c('seu','seu_all','notum'))

    message_('Starting run_numbat()')
    out = run_numbat(count_mat = count_mat, # gene x cell integer UMI count matrix
                    lambdas_ref = ref_internal, # reference expression profile, a gene x cell type normalized expression level matrix
                    df_allele = df_allele, # allele dataframe generated by pileup_and_phase script
                    out_dir = outdir,
                    genome = "hg38",
                    min_LLR = 2,
                    ncores = parallel::detectCores(),
                    plot = TRUE)

    saveRDS(out, numbat_out_rds)
    message_('Done with main Numbat process. Moving on to plots')
} else {
    message_('Main Numbat process already complete. Moving on to plots.')
}

################################################################################
# Plots
################################################################################
message_(paste('Starting plotting for sample', pat))
mypal = c('1' = 'gray', '2' = "#377EB8", '3' = "#4DAF4A", '4' = "#984EA3", 
          '5'="#ff9768", '6'='#ae1717', '7'='#0f04b5', '8'="#f87382",
          '9'='green','10'='yellow','11'='deeppink')

# Read-in numbat output
nb = Numbat$new(out_dir = outdir)
seu <- readRDS(seu_file)


# Single-cell CNV calls
cnv_calls<- nb$joint_post %>% select(cell, CHROM, seg, cnv_state, p_cnv, p_cnv_x, p_cnv_y)
write.csv(cnv_calls, file=cnv_calls_file)

# Writes to stdout but doesn't do anything
table(cnv_calls$cnv_state)
cnv_calls %>% group_by(cnv_state) %>% arrange(p_cnv,desc=F)

# Clone info
clones<-dim(table(nb$clone_post$clone_opt))
clone_info<-nb$clone_post
seu$cell<-colnames(seu)
seu@meta.data<-left_join(seu@meta.data,clone_info,by='cell')
rownames(seu@meta.data)<-colnames(seu)

# Actual plots
pdf(plots_file, width = 10)
if (T) { # Thinly veiled pretense to indent this block
    # Copy number landscape and single-cell phylogeny
    p <- nb$plot_phylo_heatmap(clone_bar = TRUE, p_min = 0.9,pal_clone = mypal)
    print(p)

    # Consensus copy number segments
    p <- nb$plot_consensus()
    print(p)

    # Bulk CNV profiles
    p <- nb$bulk_clones %>% 
        filter(n_cells > 50) %>%
        plot_bulks(min_LLR = 10, # filtering CNVs by evidence
                    legend = TRUE)
    print(p)

    # clones
    p <- DimPlot(seu, group.by = 'clone_opt',shuffle = T, cols = mypal[1:clones])
    print(p)
    p <- DimPlot(seu, group.by = 'GT_opt',shuffle = T)
    print(p)

    # Tumor versus normal probability
    p1<-FeaturePlot(seu, features  = c('p_cnv'),order = T)+
    scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, limits = c(0,1), name = 'Posterior')+
    ggtitle('Tumor vs normal probability\n(joint)')

    p1a<-FeaturePlot(seu, features  = c('p_cnv'),order = T)+
    scale_color_gradientn(colors = viridis(20))+
    ggtitle('Tumor vs normal probability\n(joint): v2')

    p2<-FeaturePlot(seu, features  = c('p_cnv_x'),order = T)+
    scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, limits = c(0,1), name = 'Posterior')+
    ggtitle('Tumor vs normal probability\n(gex)')

    p3<-FeaturePlot(seu, features  = c('p_cnv_y'),order = T)+
    scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, limits = c(0,1), name = 'Posterior')+
    ggtitle('Tumor vs normal probability\n(allele)')
    print((p1+p1a+p2 +p3)+plot_layout(ncol=2))

    # Tumor phylogeny
    p <- nb$plot_sc_tree(
    label_size = 3, 
    branch_width = 0.5, 
    tip_length = 0.5,
    pal_clone = mypal,
    tip = TRUE)
    print(p)

    # mutational history
    p <- nb$plot_mut_history(pal=mypal)
    print(p)
}
dev.off()


saveRDS(nb, numbat_clone_file)
saveRDS(seu, numbat_seu_file)
write.csv(seu@meta.data,file=numbat_seu_metadata_file)


message_(paste('Finished:', pat))